Flex APIs / NM Placement

smarella opened the pull request:
- Added "profile" parameter to flexdown API
- Added "constraint" parameter to both flexup/flexdown APIs. Currently supports:
  - only 1 constraint
  - only LIKE operator
  - hostname field
  - Mesos slave attribute field
- Improved the parameter validation of both flexup/down (Continuation of work done by @klucar)
- Changed the behavior of flexdown to pick NMs in "pending", "staging" states ahead of those in "active"
- Updated the API doc.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/api/ClustersResource.java:
```diff
@@ -100,46 +95,114 @@ public Response flexUp(FlexUpClusterRequest request) {
-        LOGGER.info("Received flexdown request");
-        LOGGER.info("Instances: " + instances);
+        String profile = request.getProfile();
+        List<String> constraints = request.getConstraints();
```
Do we assume a single constraint for now? Do we plan more constraints? If not, why do we define it as an array?

smarella replied:
We support just one constraint for now. The definition of array is to allow the possibility of multiple constraints in the future.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/api/ClustersResource.java:
```diff
@@ -100,46 +95,114 @@ public Response flexUp(FlexUpClusterRequest request) {
-    private Integer getFlexibleInstances() {
-        // this follows the logic of myriadOperations.flexDownCluster
+    private boolean validateConstraintsSize(List<String> constraints, ResponseBuilder response) {
+      if (constraints.size() > 1) {
```
If later on we will have multiple constraints how will we define relationship between them: "OR", "AND"?

smarella replied:
Yes.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java:
```diff
@@ -37,103 +37,124 @@
+                meetsConstraint(nodeTask, constraint)) {
+              this.schedulerState.makeTaskKillable(taskId);
+              numPendingTasksScaledDown++;
+              if (numPendingTasksScaledDown == numInstancesToScaleDown) {
```
I would put >= . I know ">" should not happen but just in case

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java:
```diff
@@ -37,103 +37,124 @@
-                iterator.remove();
+          for (Protos.TaskID taskId : pendingTasks) {
+            NodeTask nodeTask = schedulerState.getTask(taskId);
+            if (nodeTask.getProfile().getName().equals(profile.getName()) &&
```
nodeTask potentially could be null, as we get set of pendingTasks get prior and it can be removed meanwhile.

smarella replied:
Good catch. Will fix it.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java:
```diff
@@ -37,103 +37,124 @@
+          Set<Protos.TaskID> stagingTasks = Sets.newHashSet(this.schedulerState.getStagingTaskIds());
+
+          for (Protos.TaskID taskId : stagingTasks) {
+            NodeTask nodeTask = schedulerState.getTask(taskId);
```
Let's check for null

smarella replied:
Sure.

smarella replied:
sure.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java:
```diff
@@ -37,103 +37,124 @@
+              this.schedulerState.makeTaskKillable(taskId);
+              numPendingTasksScaledDown++;
+              if (numPendingTasksScaledDown == numInstancesToScaleDown) {
+                break;
```
Does it make more sense to call logging function here and return from the method, so not to check later on for counters < numInstancesToScaleDown

smarella replied:
Ideally (performance wise), it definitely helps. In this case, I just preferred to have the code more readable. Hence, all the counters are accumulated and printed out in LOG towards the end of the method.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/constraints/LikeConstraint.java:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  public boolean matchesHostName(String hostname) {
+    return lhs.equalsIgnoreCase("hostname") && hostname != null && Pattern.matches(rhsRegex, hostname);
```
Let's compile regex upfront in constructor - otherwise it will compile it every time here

smarella replied:
Great catch. Will fix it.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java:
```diff
@@ -37,103 +37,124 @@
+          } else {
+            return likeConstraint.matchesSlaveAttributes(nodeTask.getSlaveAttributes());
+          }
+      }
```
any break; and default: conditions?

smarella replied:
Oh yes. I think it makes sense to have a `default: return false;` to imply that the currently supported constraint is just LIKE. Will add it.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/constraints/LikeConstraint.java:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  public boolean isConstraintOnHostName() {
+    return lhs.equalsIgnoreCase("hostname");
```
Let's make "hostname" a const, even better enum, as later one you may have more constraints?

smarella replied:
Will make it a const. Not sure if enum is needed at this point.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/state/SchedulerState.java:
```diff
@@ -191,7 +192,21 @@ public synchronized void removeTask(Protos.TaskID taskId) {
+          }
+        }
+      }
+      return Collections.unmodifiableCollection(activeNodeTasks);
```
Why do you need unmodifiable - it is already new ArrayList

smarella replied:
Not sure why. I just made a new method from `getActiveTasks()` that takes `NMProfile` as parameter and filters the active tasks by the profile.
@sdaingade - do you know why we need an unmodifiable collection for all the returned objects of `SchedulerState` ?.

yufeldman commented on myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java:
```diff
@@ -74,20 +76,35 @@ public void onEvent(ResourceOffersEvent event, long sequence,
     try {
       for (Iterator<Offer> iterator = offers.iterator(); iterator.hasNext();) {
         Offer offer = iterator.next();
+        NodeTask nodeTask = schedulerState.getNodeTask(offer.getSlaveId());
```
I don't quite understand this piece of code. Did you mean to set slaveAttributes on the nodetask and use it later on?
As far as I see we setSlaveId() when task is launched (line 128). Should you set slaveAttributes at the same time and not search and set repeatedly?

smarella replied:
Yes. We need to set the slave attributes in NodeTask and use it later on (when a flexdown is requested with constraints).

I originally had this code at the place where we are launching the tasks. However, I heard Mesos is working on doing dynamic slave attributes. If the attributes on a slave change after Myriad launches the tasks, the new offers from the same slave may have a different set of attributes. I'm ok to move this down to the place where we perform launch tasks. @adam-mesos thoughts?

yufeldman replied:
If so I recommend saving slave attributes separately and update only if they were changed, otherwise if you have hundreds of tasks it will be extremely time consuming to search every time for ones that fit that slave, as offers happen every 5 sec?

smarella replied:
Are you saying setting slave attributes into NodeTask upon receiving an offer from the corresponding slave (every 5 seconds) degrades performance? I don't understand why that might be the case as a 'set' would simply update the reference count for List<Attribute>.

yufeldman replied:
This is the content of the method you are using:
public synchronized NodeTask getNodeTask(SlaveID slaveId) {
        for (Map.Entry<Protos.TaskID, NodeTask> entry : tasks.entrySet()) {
            if (entry.getValue().getSlaveId() != null &&
                entry.getValue().getSlaveId().equals(slaveId)) {
                return entry.getValue(); 
            }
        }
        return null;
    }

smarella replied:
Ok. SchedulerState.getnodeTask() and perhaps the whole SchedulerState class obviously need improvement. We are storing the tasks in various maps keyed off of TaskId, but there seems to be a need to query based on the slave id.

adam-mesos replied:
Although I don't think anybody's currently working on dynamic slave attributes, it's definitely on the roadmap.
See https://issues.apache.org/jira/browse/MESOS-1739 and https://issues.apache.org/jira/browse/MESOS-3059
When it does happen, Mesos will likely have to notify frameworks with running tasks on the reconfigured node that the attributes have changed, in case the framework would want to cancel/reschedule given the new knowledge.

Imagine a node that does not have the FUBAR attribute when we originally launch a Myriad NM task on it, but later gets tagged as FUBAR. What should the expected behavior be if we ask to flex-down a NM on a FUBAR node? Do we flex down the task because it is on a FUBAR node, or do we leave this task running, since the node was not FUBAR when we launched the task?

yufeldman replied:
In this case may be we better piggyback on that notification (Mesos will likely have to notify frameworks with running tasks on the reconfigured node that the attributes have changed) regarding attributes changes to change attributes for running tasks instead of doing it every time offer is received. Thoughts?

smarella replied:
@adam-mesos I think the behavior should be to flex down the task because it is currently on a FUBAR node, although it was originally launched when the node was not marked FUBAR. 

@yufeldman Agreed. Myriad as a framework should implement the attribute change notification from mesos.

smarella replied:
@yufeldman I realized if we update the slave attributes only during launchTasks, we may lose the slave attributes when RM fails over. This is because slave attributes in NodeTask are currently not being persisted.

We have a couple of options:
- Update slave attributes on every offer (like we are doing it now).
- Persist slave attributes with NodeTask. Implement attribute change notification (in the future) and update NodeTask when slave attributes change.

smarella commented:
I've addressed most of the review comments (dynamic slave attributes is for future). I've also fixed/cleaned up parts of flex down code to address some edge cases. This is stable AFAIK. 

Unless there are any comments/concerns, this PR is ready for merge.

smarella commented:
Merged this PR to https://github.com/apache/incubator-myriad/commits/master. 
Start commit: https://github.com/apache/incubator-myriad/commit/afa9ac904e11d72c544f6cd2499cf724dc62ab35

End Commit: https://github.com/apache/incubator-myriad/commit/e1841949196f62e861ac0a7e6f39a77d3f55fcdd

Also merging this into mesos/myriad to allow @yufeldman rebase her JHS work.
